runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -359,13 +534,100 @@ public void finishBundle() {
         emitResults();
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
+      } finally {
+        remoteBundle = null;
+      }
+      if (bundleFinishedCallback != null) {
+        bundleFinishedCallback.run();
+        bundleFinishedCallback = null;
       }
     }
 
+    /** Key for timer which has not been registered yet. */
+    Object getTimerKeyForRegistration() {
+      return keyForTimerToBeSet;
+    }
+
+    /** Key for timer which is about to be fired. */
+    void setTimerKeyForFire(Object key) {
+      this.keyForTimerToBeFired = key;
+    }
+
+    boolean isBundleInProgress() {
+      return remoteBundle != null;
+    }
+
+    void setBundleFinishedCallback(Runnable callback) {
+      this.bundleFinishedCallback = callback;
+    }
+
     private void emitResults() {
       KV<String, OutputT> result;
       while ((result = outputQueue.poll()) != null) {
-        outputManager.output(outputMap.get(result.getKey()), (WindowedValue) result.getValue());
+        final String outputPCollectionId = Preconditions.checkNotNull(result.getKey());
+        TupleTag<?> tag = outputMap.get(outputPCollectionId);
+        WindowedValue windowedValue =
+            Preconditions.checkNotNull(
+                (WindowedValue) result.getValue(),
+                "Received a null value from the SDK harness for %s",
+                outputPCollectionId);
+        if (tag != null) {
+          // process regular elements
+          outputManager.output(tag, windowedValue);
+        } else {
+          final String timerCollectionId = extractTimerPCollectionId(outputPCollectionId);
+          TimerSpec timerSpec = extractTimerSpec(timerCollectionId);
+          Timer timer =
+              Preconditions.checkNotNull(
+                  (Timer) ((KV) windowedValue.getValue()).getValue(),
+                  "Received null Timer from SDK harness: %s",
+                  windowedValue);
+          LOG.debug("Timer received: {} {}", outputPCollectionId, timer);
+          for (Object window : windowedValue.getWindows()) {
+            StateNamespace namespace = StateNamespaces.window(windowCoder, (BoundedWindow) window);
+            TimerInternals.TimerData timerData =
+                TimerInternals.TimerData.of(
+                    timerCollectionId, namespace, timer.getTimestamp(), timerSpec.getTimeDomain());
```

Comment:
The notion "timerCollectionId" is ambiguous, as timers are associated with two distinct PCollections at execution time. 

Timers are absolutely determined by their transform name and local name; probably it makes sense to use that (let's call it fullyQualifiedTimerName) for the TimerData here, and then also have a mapping fullyQualifiedTimerName -> timerInputPCollectionId for plumbing timers to the SDK, and a map timerOutputPCollectionId -> fullyQualifiedTimerName for recording timers from the SDK (like we do here).

Reply:
>Timers are absolutely determined by their transform name and local name; probably it makes sense to use that

That's what is used here. The keys of `timerReferenceMap` are transform name + local name. The "extractTimerPCollectionId" method does the mapping of output collection to timer name.

>and then also have a mapping fullyQualifiedTimerName -> timerInputPCollectionId for plumbing timers to the SDK, and a map timerOutputPCollectionId -> fullyQualifiedTimerName for recording timers from the SDK (like we do here).

I couldn't find a nice way to create a map timerOutputPCollectionId -> fullyQualifiedTimerName.

Reply:
I think it will be best we merge this PR and revert #6986.  BEAM-5999 can then rebase and address this issue.

Reply:
OK, the problem is that there are multiple copies of components and pcollections floating around...

The right way to solve this is to patch in https://github.com/mxm/beam/pull/1/commits/55baf15befea582a3fc7796f9c0a4a031c3485b3 and then use stageBundleFactory.getProcessBundleDescriptor().getTimerSpecs() in SdkHarnessDoFnRunner to construct timerReferenceMap or whatever else is needed, which has the full information.

Reply:
https://github.com/mxm/beam/pull/1 This also avoids hacks to guess the output name from the input name.

Reply:
Neat. With your fix I was able to build a `timerOutputIdToSpecMap` from output collection id to TimerSpec. This also got rid of some boiler plate code which I previously used to extract the TimerSpec.